### PR TITLE
Binding factory to contract

### PR DESCRIPTION
**Pull request by Soren025**
I ran into a situation where I needed to pair a factory with the type of object it was created with. I had thought I had found a solution, but discovered that you are currently unable to bind factories to contract like normal types. Looking into the `DiContainer` I noticed this would be a simple implementation, so I decided to make a pull request.

Below is a simplified example of how I intend to use such a feature...

```csharp
public abstract class Foo
{

}

public interface IFooFactory : IFactory<Foo>
{
    Type FooType { get; }
}

public sealed class FooFactory<TFoo> : Factory<TFoo>, IFooFactory
    where TFoo : Foo
{
    public Type FooType { get { return typeof(TFoo); } }

    public new TFoo Create()
    {
        base.Create();
    }

    // Explicit implementation of `IFactory<Foo>` from `IFooFactory`
    Foo IFactory<Foo>.Create()
    {
        Create();
    }
}

public class FooManager
{
    private Dictionary<Type, IFooFactory> fooFactories;

    [Inject]
    public FooManager(List<IFooFactory> fooFactoryList)
    {
        fooFactories = new Dictionary<Type, IFooFactory>(fooFactoryList.Count);
        foreach (var fooFactory in fooFactoryList)
        {
            fooFactories[fooFactory.FooType] = fooFactory;
        }
    }

    // Methods to create `Foo` types
}

public abstract class FooInstallerBase : MonoInstaller
{
    protected abstract void BindFooFactory<TFoo>()
        where TFoo : Foo
    {
        Container.BindFactoryContract<Foo, IFooFactory, FooFactory<TFoo>>()
            .WhenInjectedInto<FooManager>();
    }
}
```

Basically the idea behind my example is to selectively bind types of `Foo` into factories based on what installers that extend `FooInstallerBase` get installed. Then I can dynamically inject them all as non generic interfaces in list form to be processed into a `Dictionary<Type, IFooFactory>` for management by the `FooManager`.

### CODE REVIEW

**Comment by svermeulen** [comment]
Yeah, I can see how it'd be useful for cases where you have your own factory interface/base class.  Thanks!
